Refactored code to have separate CLI and logic

eirannejad opened the pull request:
- rhino-agnostic cli builder and logic module
- added build script for rhino
- multiple targets

eirannejad committed:
Refactored code to have separate CLI and logic

- rhino-agnostic cli builder and logic module
- added build script for rhino
- multiple targets

eirannejad commented:
- Refactored the code to separate CLI from logic. The logic dlls are now used with pyRevit on Revit to generate stubs from the running instance of Revit and for that specific API version
- Added x86 and x64 targets for both the cli and the logic module
- Added powershell build script for default Rhino stubs
- Updated .gitignore

CLI usage
```
Usage:
    pystubsbuilder (-h | --help)
    pystubsbuilder (-V | --version)
    pystubsbuilder [--dest=<dest_path>] [--search=<search_path>...] <target_dll>...
```

Accepts destination path `--dest`
Accepts multiple search paths `--search`
Accepts multiple target dlls

eirannejad committed:
Removed x86 builds and tested Rhino stubs build

sbaer commented:
If I run `build-rhino-stubs.ps1`, the resulting stubs end up with a slightly different path. I would fix this, but I don't want to mess up expected paths for your Revit project

eirannejad commented:
Whoops you are right. I'll revise to support both methods. Great catch

eirannejad committed:
Added more options to control the output dir structure

    --prefix=<prefix>           Root namespace directory prefix
    --postfix=<postfix>         Root namespace directory postfix
    --dest-is-root              Use destination path for root namespace

eirannejad commented:
Added more options to control the output dir structure

```
    --prefix=<prefix>           Root namespace directory prefix
    --postfix=<postfix>         Root namespace directory postfix
    --dest-is-root              Use destination path for root namespace
```

`--prefix` and `--postfix` can extend the final root namespace directory name.
`--dest-is-root` is to force the utility to use the `--dest` path as the root namespace directory

sbaer commented:
Hmm.. the paths are correct, but for some reason the generated .pyi files are being created with contents in a different order. That may be fine, but I need more time to review

sbaer commented:
Something is definitely different in the stubs generation logic. If I roll back to master, I can see that the stubs are getting generated the same way they used to be getting generated.

eirannejad commented:
Seems like the order is different. I didn't mess with the stub generator code. I think we need to add ordering for consistent output

sbaer commented on builder/PythonStubsBuilder.cs:
```diff
@@ -0,0 +1,283 @@
+                cfgs = new PythonStubsBuilderConfigs();
+
+            // prepare resolver
+            AppDomain.CurrentDomain.AssemblyResolve += AssemblyResolve; ;
```
We're adding the same AssemblyResolver over and over every time this function is called.

eirannejad replied:
Should I add a static bool flag to mark "AlreadyHandling" or there is a better way?

```csharp
if (!alreadyHandling)
    AppDomain.CurrentDomain.AssemblyResolve += AssemblyResolve;
```

sbaer replied:
You could also just always do a `-=` before a `+=` call
